File: apps/host/internal/githost/clone.go
package githost

import (
	"context"
	"encoding/json"
	"errors"
	"strings"

	pb "armadra.local/host/gen/armadra/v1"
	"armadra.local/host/internal/storage"
	"google.golang.org/protobuf/proto"
)

// Clone jobs (Go Host 业务所有权迁移 §2.8, §3.1 "URL 只存摘要与展示用去凭据形式").
//
// A clone is not an operation on a repository, because the repository does not
// exist yet: there is nothing to lock, nothing to state a precondition against,
// and no worktree to serialize with. So it is its own record with its own
// lifecycle, and it never enters the queue.
//
// The URL is the reason this file exists at all. A clone URL can carry a
// credential — `https://user:token@host/repo.git` is a shape people paste — and
// a row that kept one would turn a database read into a credential leak, on a
// machine whose whole point is that other devices reach it. So the URL is
// forwarded to the execution host, which needs it, and what is *stored* is the
// digest, which is enough to recognise the same clone twice, plus the redacted
// display form produced by the side that held the original.

// MaxURLBytes bounds a clone URL. A repository address is a URL, not a payload.
const MaxURLBytes = 2048

// cloneRequest is the version-locked body the Worker's own clone route takes.
// It is assembled here rather than accepted from the caller so the URL this
// Host forwards is the one it validated.
type cloneRequest struct {
	URL        string `json:"url"`
	TargetPath string `json:"targetPath,omitempty"`
}

// cloneOutcome is the Worker's answer: the job it started, the redacted URL it
// produced from the original, and how far it has got.
type cloneOutcome struct {
	JobID      string `json:"jobId"`
	DisplayURL string `json:"displayUrl"`
	TargetPath string `json:"targetPath"`
	Progress   uint32 `json:"progress"`
	State      string `json:"state"`
	Message    string `json:"messageCode"`
}

func (s *Service) cloneKey(workspaceID, jobID string) storage.Key {
	return storage.Key{Kind: CloneKind, ID: jobID, WorkspaceID: workspaceID}
}

func (s *Service) cloneJob(ctx context.Context, workspaceID, jobID string) (*pb.GitCloneJob, error) {
	if !validID(jobID) {
		return nil, ErrInvalid
	}
	entity, err := s.store.Read(ctx, s.cloneKey(workspaceID, jobID))
	if errors.Is(err, storage.ErrNotFound) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, err
	}
	if entity.Deleted {
		return nil, ErrNotFound
	}
	return decodeClone(entity)
}

func (s *Service) putClone(ctx context.Context, operationID, workspaceID string, job *pb.GitCloneJob, expected uint64) (*pb.GitCloneJob, storage.ApplyResult, error) {
	encoded, err := clonePayload(job)
	if err != nil {
		return nil, storage.ApplyResult{}, err
	}
	result, err := s.store.Apply(ctx, operationID, []storage.Change{{
		Key:              s.cloneKey(workspaceID, job.GetJobId()),
		ExpectedRevision: expected,
		Payload:          encoded,
	}})
	if err != nil {
		return nil, storage.ApplyResult{}, err
	}
	stored, err := s.cloneJob(ctx, workspaceID, job.GetJobId())
	if err != nil {
		return nil, storage.ApplyResult{}, err
	}
	return stored, result, nil
}

// validCloneURL is the shape check this Host makes before forwarding. It is
// deliberately not an allowlist of hosts: which remotes a person may clone is
// the execution host's own configuration, and duplicating it here would put two
// policies in disagreement. What is refused is a URL that is not one.
func validCloneURL(value string) bool {
	if value == "" || len(value) > MaxURLBytes || strings.ContainsAny(value, "\x00\n\r") {
		return false
	}
	for _, prefix := range []string{"https://", "http://", "ssh://", "git://", "file://", "git@"} {
		if strings.HasPrefix(value, prefix) {
			return true
		}
	}
	// A bare local path is a legitimate clone source and is how a bare mirror
	// on the same machine is addressed.
	return validPath(value)
}

func cloneState(value string) pb.GitCloneState {
	switch value {
	case "running":
		return pb.GitCloneState_GIT_CLONE_STATE_RUNNING
	case "succeeded":
		return pb.GitCloneState_GIT_CLONE_STATE_SUCCEEDED
	case "cancelled":
		return pb.GitCloneState_GIT_CLONE_STATE_CANCELLED
	case "failed":
		return pb.GitCloneState_GIT_CLONE_STATE_FAILED
	default:
		return pb.GitCloneState_GIT_CLONE_STATE_UNSPECIFIED
	}
}

// forwardClone sends one clone request to the execution host and decodes its
// answer. The `GitRead` frame carries it because a clone is a forwarded
// operation with no queue behind it, exactly like a repository query.
func (s *Service) forwardClone(ctx context.Context, method pb.GitReadMethod, workspaceID, root string, body any) (cloneOutcome, error) {
	var outcome cloneOutcome
	encoded, err := json.Marshal(body)
	if err != nil {
		return outcome, err
	}
	result, err := s.executor.ReadGit(ctx, &pb.GitRead{
		Scope:         &pb.RepositoryScope{WorkspaceId: workspaceID},
		RequestJson:   encoded,
		WorkspaceRoot: root,
		Method:        method,
	})
	if err != nil {
		return outcome, err
	}
	if result == nil || result.GetHttpStatus() < 200 || result.GetHttpStatus() >= 300 {
		return outcome, ErrInvalid
	}
	if err = json.Unmarshal(result.GetResponseJson(), &outcome); err != nil {
		return outcome, ErrInvalid
	}
	if outcome.JobID == "" || !validID(outcome.JobID) {
		return outcome, ErrInvalid
	}
	return outcome, nil
}

// replayClone answers a repeated clone with the job the first one started.
func (s *Service) replayClone(ctx context.Context, workspaceID, operationID string) (*pb.StartGitCloneResponse, bool, error) {
	if !textOperationID(operationID) {
		return nil, false, ErrInvalid
	}
	result, found, err := s.store.Receipt(ctx, operationID)
	if err != nil || !found {
		return nil, false, err
	}
	for _, revision := range result.Revisions {
		if revision.Kind != CloneKind || revision.WorkspaceID != workspaceID {
			continue
		}
		job, err := s.cloneJob(ctx, workspaceID, revision.ID)
		if err != nil {
			return nil, false, err
		}
		result.Replayed = true
		return &pb.StartGitCloneResponse{Job: job, Receipt: receipt(result)}, true, nil
	}
	return nil, false, nil
}

// Clone starts one clone on the execution host and records it.
func (s *Service) Clone(ctx context.Context, caller Caller, request *pb.StartGitCloneRequest) (*pb.StartGitCloneResponse, error) {
	if err := s.authorizeWrite(ctx, caller); err != nil {
		return nil, err
	}
	url := request.GetUrl()
	if !validCloneURL(url) {
		return nil, ErrInvalid
	}
	if target := request.GetTargetPath(); target != "" && len(target) > MaxPathBytes {
		return nil, ErrInvalid
	}
	root, err := s.workspaceRoot(ctx, caller.WorkspaceID)
	if err != nil {
		return nil, err
	}
	// A retry is answered before anything is started. The job identifier comes
	// from the execution host, so forwarding a repeated request would start a
	// second clone into the same directory and only then discover that the
	// idempotency key had already been used.
	if replay, found, err := s.replayClone(ctx, caller.WorkspaceID, request.GetOperationId()); err != nil {
		return nil, err
	} else if found {
		return replay, nil
	}
	outcome, err := s.forwardClone(ctx, pb.GitReadMethod_GIT_READ_METHOD_CLONE_START, caller.WorkspaceID, root, cloneRequest{URL: url, TargetPath: request.GetTargetPath()})
	if err != nil {
		return nil, err
	}
	now := s.now()
	job := &pb.GitCloneJob{
		JobId:       outcome.JobID,
		WorkspaceId: caller.WorkspaceID,
		// The URL itself is never stored. Only its digest, which recognises the
		// same clone twice, and the redacted form the Worker produced.
		UrlSha256:       digest([]byte(url)),
		DisplayUrl:      outcome.DisplayURL,
		TargetPath:      outcome.TargetPath,
		Progress:        outcome.Progress,
		State:           pb.GitCloneState_GIT_CLONE_STATE_RUNNING,
		CreatedAtUnixMs: now,
		UpdatedAtUnixMs: now,
	}
	stored, result, err := s.putClone(ctx, request.GetOperationId(), caller.WorkspaceID, job, 0)
	if err != nil {
		return nil, err
	}
	return &pb.StartGitCloneResponse{Job: stored, Receipt: receipt(result)}, nil
}

// GetClone answers with the Host's record, refreshed from the execution host
// while the job is still running.
//
// A finished job is answered from the record alone. The execution host forgets
// a clone when its process ends; the Host does not, which is the point of
// keeping the row at all.
func (s *Service) GetClone(ctx context.Context, caller Caller, request *pb.GetGitCloneRequest) (*pb.GetGitCloneResponse, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	job, err := s.cloneJob(ctx, caller.WorkspaceID, request.GetJobId())
	if err != nil {
		return nil, err
	}
	if job.GetState() != pb.GitCloneState_GIT_CLONE_STATE_RUNNING || s.executor == nil {
		return &pb.GetGitCloneResponse{Job: job}, nil
	}
	root, err := s.workspaceRoot(ctx, caller.WorkspaceID)
	if err != nil {
		return &pb.GetGitCloneResponse{Job: job}, nil
	}
	outcome, err := s.forwardClone(ctx, pb.GitReadMethod_GIT_READ_METHOD_CLONE_STATUS, caller.WorkspaceID, root, map[string]string{"jobId": job.GetJobId()})
	if err != nil {
		// An unreachable execution host does not change what the Host knows.
		// The row keeps saying "running", which is the last thing anybody
		// actually observed.
		return &pb.GetGitCloneResponse{Job: job}, nil
	}
	next := proto.Clone(job).(*pb.GitCloneJob)
	next.Progress = outcome.Progress
	next.State = cloneState(outcome.State)
	next.MessageCode = outcome.Message
	next.UpdatedAtUnixMs = s.now()
	if next.GetState() == pb.GitCloneState_GIT_CLONE_STATE_UNSPECIFIED {
		next.State = job.GetState()
	}
	if next.GetState() == job.GetState() && next.GetProgress() == job.GetProgress() {
		return &pb.GetGitCloneResponse{Job: job}, nil
	}
	stored, _, err := s.putClone(ctx, "githost/"+caller.WorkspaceID+"/clone/"+job.GetJobId()+"/"+outcome.State, caller.WorkspaceID, next, job.GetRevision())
	if err != nil {
		return &pb.GetGitCloneResponse{Job: job}, nil
	}
	return &pb.GetGitCloneResponse{Job: stored}, nil
}

// CancelClone stops a running clone. Unlike a push there is no unknown outcome
// to worry about: a clone writes into a directory this side named, and
// cancelling it cleans up only what this clone created (Git 设计 §3).
func (s *Service) CancelClone(ctx context.Context, caller Caller, request *pb.CancelGitCloneRequest) (*pb.CancelGitCloneResponse, error) {
	if err := s.authorizeWrite(ctx, caller); err != nil {
		return nil, err
	}
	job, err := s.cloneJob(ctx, caller.WorkspaceID, request.GetJobId())
	if err != nil {
		return nil, err
	}
	if job.GetState() != pb.GitCloneState_GIT_CLONE_STATE_RUNNING {
		return nil, ErrTerminal
	}
	root, err := s.workspaceRoot(ctx, caller.WorkspaceID)
	if err != nil {
		return nil, err
	}
	if _, err = s.forwardClone(ctx, pb.GitReadMethod_GIT_READ_METHOD_CLONE_CANCEL, caller.WorkspaceID, root, map[string]string{"jobId": job.GetJobId()}); err != nil {
		return nil, err
	}
	next := proto.Clone(job).(*pb.GitCloneJob)
	next.State = pb.GitCloneState_GIT_CLONE_STATE_CANCELLED
	next.MessageCode = "git.clone.cancelled"
	next.UpdatedAtUnixMs = s.now()
	stored, result, err := s.putClone(ctx, request.GetOperationId(), caller.WorkspaceID, next, job.GetRevision())
	if err != nil {
		return nil, err
	}
	return &pb.CancelGitCloneResponse{Job: stored, Receipt: receipt(result)}, nil
}

// activeClones counts the clones this Host still believes are running. A switch
// reads it: moving the domain while a clone is in flight would leave the only
// record of a half-written directory in a process that is about to stop being
// the writer.
func (s *Service) activeClones(ctx context.Context, workspaceID string) (int, error) {
	count := 0
	after := ""
	for {
		page, err := s.store.List(ctx, storage.ListOptions{WorkspaceID: workspaceID, Kind: CloneKind, AfterID: after, Limit: MaxPage})
		if err != nil {
			return 0, err
		}
		for _, entity := range page.Entities {
			job, err := decodeClone(entity)
			if err != nil {
				return 0, err
			}
			if job.GetState() == pb.GitCloneState_GIT_CLONE_STATE_RUNNING {
				count++
			}
		}
		if !page.HasMore {
			return count, nil
		}
		after = page.NextID
	}
}

File: apps/runtime/src/worker/git/clone.rs
//! Clones on the Worker channel (Git 设计 §3 Init/Clone, business migration
//! §2.8).
//!
//! A clone is the one thing in this domain with no repository behind it. There
//! is nothing to lock, no worktree to serialize against and no precondition to
//! have been decided against, which is why it travels on the read channel and
//! never enters the Host's queue.
//!
//! What it *does* have is a lifetime longer than the frame that starts it. The
//! job lives in this process's registry and its `git` child keeps running after
//! the answer has been written, so the Host has to hold this Worker open to
//! poll it — a resident clone Worker, which is exactly what
//! `cmd/armadra-host/git.go` keeps for these three methods. A Worker started
//! per operation would begin a clone that no later process could reach, which
//! is why this used to answer UNSUPPORTED.
//!
//! The URL is the other thing worth stating. It arrives here and goes no
//! further: what leaves is the redacted form, produced by this side because
//! this is the side that held the original. A `https://user:token@host/repo`
//! that reached the Host's database would turn a row read into a credential
//! leak.

use std::path::{Path, PathBuf};

use armadra_protocol::v1::{GitOperationState, GitReadResult};
use serde::{Deserialize, Serialize};

use super::{invalid, progress, result};
use crate::{
    error::{AppError, AppResult},
    git,
    security::resolve_in_root,
};

/// What the Host sends to start one. It is the same version-locked JSON every
/// other forwarded read uses.
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct StartBody {
    url: String,
    /// Where the clone should land, workspace-relative. Empty lets the
    /// directory name be derived from the URL, which is what the dialog does
    /// when a person leaves the field alone.
    #[serde(default)]
    target_path: Option<String>,
}

/// The answer, in the shape `githost/clone.go` decodes.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct Outcome {
    job_id: String,
    /// Credentials already removed, by the side that held them.
    display_url: String,
    target_path: String,
    progress: u32,
    /// `running`, `succeeded`, `failed` or `cancelled`.
    state: &'static str,
    message_code: String,
}

fn answer(outcome: &Outcome) -> AppResult<GitReadResult> {
    let body = serde_json::to_vec(outcome)
        .map_err(|_| AppError::Internal("The clone answer could not be encoded".into()))?;
    Ok(result(200, body))
}

/// The three terminal words the Host's own enum uses, plus `running`.
///
/// A cancelled clone is not a failed one: the directory it was writing was left
/// where it is on purpose, and telling a person "the clone failed" would send
/// them looking for a reason that does not exist.
fn state_of(status: &git::CloneStatus) -> &'static str {
    match status.state {
        git::CloneState::Running => "running",
        git::CloneState::Done => "succeeded",
        git::CloneState::Error if status.cancelled => "cancelled",
        git::CloneState::Error => "failed",
    }
}

fn message_of(status: &git::CloneStatus) -> String {
    match status.state {
        git::CloneState::Running => "git.clone.running".into(),
        git::CloneState::Done => "git.clone.succeeded".into(),
        git::CloneState::Error if status.cancelled => "git.clone.cancelled".into(),
        git::CloneState::Error => "git.clone.failed".into(),
    }
}

/// The clone's destination, resolved inside the registered workspace root.
///
/// The Host said where the files are; this side decides that the target is
/// under it. A caller-supplied absolute path is refused rather than resolved,
/// which is the same rule every other git frame follows: a root is a claim, not
/// a permission.
fn destination(workspace: &Path, target: Option<&str>) -> AppResult<(PathBuf, Option<String>)> {
    let requested = target.map(str::trim).unwrap_or_default();
    if requested.is_empty() {
        // No destination named: the parent is the workspace root and the
        // directory name comes from the URL, which is what the dialog does.
        return Ok((resolve_in_root(workspace, ".")?, None));
    }
    if Path::new(requested).is_absolute() {
        return Err(AppError::BadRequest(
            "A clone destination is workspace-relative".into(),
        ));
    }
    let (parent, name) = match requested.trim_end_matches('/').rsplit_once('/') {
        Some((parent, name)) => (parent, name),
        None => (".", requested),
    };
    if name.is_empty() {
        return Err(AppError::BadRequest(
            "A clone destination names no directory".into(),
        ));
    }
    // Only the parent is resolved: the destination itself must not exist yet,
    // and `start_clone` is what creates it after checking that.
    Ok((resolve_in_root(workspace, parent)?, Some(name.to_owned())))
}

/// The source, as this side will hand it to `git clone`.
///
/// A remote URL goes through the Runtime's own allowlist unchanged: `https://`,
/// `ssh://` and the scp-like form, and nothing else — `ext::` and friends are
/// refused there for the reason they have always been refused.
///
/// A **local** source is the one thing this path allows that the HTTP route
/// does not, and the difference is the registered root. The HTTP route clones
/// into a parent directory the caller names, so a local source there would be
/// one arbitrary path copied to another. Here both ends are inside the root the
/// filesystem domain registered for this workspace, which is what makes cloning
/// a bare mirror that sits beside the project — the ordinary way to work
/// offline, and the shape a `git worktree`-style layout already has — a
/// bounded operation rather than an open one.
fn source(workspace: &Path, url: &str) -> AppResult<String> {
    if let Ok(remote) = git::validate_clone_url(url) {
        return Ok(remote);
    }
    let candidate = if Path::new(url).is_absolute() {
        let resolved = std::fs::canonicalize(url)
            .map_err(|_| AppError::BadRequest("Repository URL is invalid".into()))?;
        let root = workspace
            .canonicalize()
            .unwrap_or_else(|_| workspace.to_path_buf());
        if !resolved.starts_with(&root) {
            return Err(AppError::Forbidden(
                "A local clone source must be inside the workspace".into(),
            ));
        }
        resolved
    } else {
        resolve_in_root(workspace, url)?
    };
    if !candidate.is_dir() {
        return Err(AppError::BadRequest("Repository URL is invalid".into()));
    }
    candidate
        .to_str()
        .map(str::to_owned)
        .ok_or_else(|| AppError::BadRequest("Clone source must be a UTF-8 path".into()))
}

/// Starts one clone and answers with the job the Host will poll.
pub async fn start(workspace: &Path, request_json: &[u8]) -> AppResult<GitReadResult> {
    let body: StartBody = serde_json::from_slice(request_json)
        .map_err(|error| invalid(&format!("the clone body does not parse: {error}")))?;
    let (parent, name) = destination(workspace, body.target_path.as_deref())?;
    let parent = parent
        .to_str()
        .ok_or_else(|| AppError::BadRequest("Clone destination must be a UTF-8 path".into()))?
        .to_owned();
    let url = source(workspace, body.url.trim())?;
    let started =
        tokio::task::spawn_blocking(move || git::start_clone_from(&url, &parent, name.as_deref()))
            .await
            .map_err(|_| AppError::Internal("The clone could not be started".into()))??;
    // The status is read back rather than assembled from the request: the
    // redacted URL and the resolved target are facts this process produced, and
    // reporting the request's own values would report what was asked for
    // instead of what happened.
    let status = git::clone_status(&started.job_id)?;
    watch(started.job_id.clone());
    answer(&Outcome {
        job_id: started.job_id,
        display_url: status.display_url.clone(),
        target_path: status.target.to_string_lossy().into_owned(),
        progress: status.percent,
        state: state_of(&status),
        message_code: message_of(&status),
    })
}

/// One clone's current state.
pub fn status(job_id: &str) -> AppResult<GitReadResult> {
    let status = git::clone_status(job_id)?;
    answer(&Outcome {
        job_id: job_id.to_owned(),
        display_url: status.display_url.clone(),
        target_path: status.target.to_string_lossy().into_owned(),
        progress: status.percent,
        state: state_of(&status),
        message_code: message_of(&status),
    })
}

/// Asks the clone to stop. The partial destination is kept: it is a directory a
/// person named, and deleting it on the way out of a cancellation is how a
/// cancel becomes a data loss.
pub fn cancel(job_id: &str) -> AppResult<GitReadResult> {
    git::cancel_clone(job_id)?;
    status(job_id)
}

/// Watches one job to its end so the Host learns it finished without polling.
///
/// It is a supplement to `GetClone`, never a replacement: the Host still reads
/// the job when it is asked to, because an upcall that never arrives must cost
/// a slower answer rather than a job that appears to run forever.
fn watch(job_id: String) {
    let Ok(handle) = tokio::runtime::Handle::try_current() else {
        return;
    };
    handle.spawn(async move {
        loop {
            tokio::time::sleep(std::time::Duration::from_millis(250)).await;
            let Ok(status) = git::clone_status(&job_id) else {
                return;
            };
            if status.state == git::CloneState::Running {
                continue;
            }
            progress::clone_finished(
                &job_id,
                &status.target.to_string_lossy(),
                match status.state {
                    git::CloneState::Done => GitOperationState::Succeeded,
                    _ if status.cancelled => GitOperationState::Cancelled,
                    _ => GitOperationState::Failed,
                },
                &message_of(&status),
            );
            return;
        }
    });
}
